app/models/ems_refresh/save_inventory_container.rb

```diff
@@ -16,20 +31,83 @@ def save_ems_container_inventory(ems, hashes, target = nil)
     ems.save!
   end
 
-  def save_container_projects_inventory(ems, hashes, target = nil)
-    return if hashes.nil?
-    target = ems if target.nil?
-
-    ems.container_projects.reset
-    deletes = if target.kind_of?(ExtManagementSystem)
-                :use_association
-              else
-                []
-              end
+  def initialize_inventory_collections(ems)
+    @inv_collections = {}
+    @inv_collections[:container_projects] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerProject,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_projects
+    )
+    @inv_collections[:container_quotas] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerQuota,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_quotas,
+      #:arel => ContainerQuota.joins(:container_project).where(:container_projects => {:ems_id => ems.id}),
+    )
+    @inv_collections[:container_quota_items] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerQuotaItem,
+      :parent => ems,
+      :association => :container_quota_items,
+      #:arel => ContainerQuotaItem.joins(:container_quota => :container_project).where(:container_projects => {:ems_id => ems.id}),
+      :manager_ref => [:container_quota, :resource],
+    )
+    @inv_collections[:container_nodes] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerNode,
+      :parent => ems,
+      :builder_params => {:ems_id => ems.id},
+      :association => :container_nodes,
+    )
+    # ContainerCondition is polymorphic child of ContainerNode & ContainerGroup
+    @inv_collections[:container_node_conditions] = ::ManagerRefresh::InventoryCollection.new(
+      :model_class => ContainerCondition,
+      :parent => ems,
+      :association => :container_node_conditions,
+      #:arel => ContainerCondition.joins(
+      #  'INNER JOIN "container_nodes" ON "container_conditions"."container_entity_id" = "container_nodes"."id"'
+      #).where(:container_entity_type => 'ContainerNode',
+      #        :container_nodes => {:ems_id => ems.id}),
+      :manager_ref => [:container_entity, :name],
+    )
+
+    # polymorphic child of ContainerNode & ContainerImage
+    # TODO these are dumb, find way to share same InventoryCollections via both.
```

Comment:
right, so it's either ems_ref, or some combination that is unique under Ems. I agree that these artificial ems_refs might be weird. Though rails allow easier queries on single ems_ref than multiple.

So given ComputerSystem, the index could be [:ems_id, :managed_entity_type, managed_entity_id]. I assume that the Node is the managed_entity here? The for targeted, you could limit it with a query. 

so it will be something like select me Ems ComputerSystems, where (managed_entity_type = 'Node' AND managed_entity_id IN (SELECT id FROM nodes where ems_ref IN (node_ems_refs))

we can generate such query, for many classes, but in general this is the downside of polymorphic, since you can't do a simple join.

------

so probably better example will be the labels, since that is has_many. We can either generate ^ large query every time (based on the targets available). Or we can do the fake ems_ref like <resource_ems_ref>_<section>_<name>, if the ems_ref of the resource is unique among all data of the ems. The you can leverage (WHERE ems_ref LIKE "<resource_ems_ref>_%" OR  LIKE "<resource_ems_ref>_%")  to get all labels of many resources across many models. Or this syntax 'where value  like any (array['%foo%', '%bar%', '%baz%']'

And there are probably more ways, many of them still effective while refreshing many thousands records at once, it the db is well indexed. 

We will have this fun with queries once we are at Targeted/Subgraph refresh. :-)